### PR TITLE
feat: Add batch timestamps

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description

Add timestamp for creation time of batches as this may deviate from execution time start and scheduled time (eventually).
Add timestamp for published time (of integration events) of completed batches for deeper insights.

Moved and made these generic in order to be able to reuse across modules
image

## References

* #972

### CODE REVIEW

File: source/dotnet/wholesale-api/IntegrationEventPublishing/IntegrationEventPublishing.Application/BatchAggregate/CompletedBatch.cs [hunk 1]
```diff
@@ -22,23 +22,23 @@ public sealed class CompletedBatch
...
         PeriodEnd = periodEnd;
         CompletedTime = completedTime;
-        IsPublished = isPublished;
     }
 
-    public bool IsPublished { get; set; }
+    /// <summary>
+    /// The time when integration events for the batch were published.
+    /// </summary>
+    public Instant? PublishedTime { get; set; }
```
Comment (lasrinnil) [inline source/dotnet/wholesale-api/IntegrationEventPublishing/IntegrationEventPublishing.Application/BatchAggregate/CompletedBatch.cs:41]: Untested?

**Review by lasrinnil — APPROVED** [review]

File: source/dotnet/wholesale-api/IntegrationEventPublishing/IntegrationEventPublishing.Application/BatchAggregate/CompletedBatch.cs (hunk 1 above)
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/IntegrationEventPublishing/IntegrationEventPublishing.Application/BatchAggregate/CompletedBatch.cs:41]: Should be tested in CompletedBatchTest. Don't you agree?

File: source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs [hunk 2]
```diff
@@ -103,6 +105,8 @@ private Batch()
 
     public Instant? ExecutionTimeStart { get; }
 
+    public Instant CreatedTime { get; }
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs:108]: Should be tested in BatchRepositoryTests. Don't you agree?

File: source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs [hunk 3]
```diff
@@ -24,6 +24,7 @@ public class Batch
     private readonly List<GridAreaCode> _gridAreaCodes;
 
     public Batch(
+        Instant createdTime,
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs:27]: So you'd prefer to pass an `IClock`. That's not what we have been doing so far.

File: source/dotnet/database-migration/Scripts/202305260800_Add_Batch_CreatedTime.sql [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿ALTER TABLE batches.Batch
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/dotnet/database-migration/Scripts/202305260800_Add_Batch_CreatedTime.sql:1]: I don't see why not. And tests are green 😄

File: source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs:108]: Just a GitHub annotation said it was not covered

File: source/dotnet/wholesale-api/IntegrationEventPublishing/IntegrationEventPublishing.Application/BatchAggregate/CompletedBatch.cs (hunk 1 above)
Comment (lasrinnil) [reply source/dotnet/wholesale-api/IntegrationEventPublishing/IntegrationEventPublishing.Application/BatchAggregate/CompletedBatch.cs:41]: Just a GitHub annotation said it was not covered

File: source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/dotnet/wholesale-api/Batches/Batches.Application/BatchAggregate/Batch.cs:27]: Nah - let's not go down that route :)

**Pull request merged**